### PR TITLE
Run runner tests in parallel

### PR DESCRIPTION
**Pull request by adombeck**
On my system, this reduces the run time of the tests from ~18s to ~5s.

Note that the downside of this is that the logs are a lot less readable, because all the tests print to stderr at the same time, so its very hard to figure out which output belongs to which tests, making debugging more difficult.

### CODE REVIEW

**Comment by jochil** [comment]
> On my system, this reduces the run time of the tests from ~18s to ~5s.
> 
> Note that the downside of this is that the logs are a lot less readable, because all the tests print to stderr at the same time, so its very hard to figure out which output belongs to which tests, making debugging more difficult.

I think it was more important to fix the underlying issue than actually run the tests in parallel... with this short runtimes I would prefer having a readable log

**Comment by fmeum** [comment]
Can't we always pass something like `-p 1 -parallel 1` to run the tests serially if we want to debug them? That way we could have both: fast and debuggable tests.

**Comment by jochil** [comment]
> Can't we always pass something like `-p 1 -parallel 1` to run the tests serially if we want to debug them? That way we could have both: fast and debuggable tests.

But if the error is not reproducible on your machine (and you got the log from CI or a user) it would be better if the logs are looking good in the first place

File: pkg/runner/libfuzzer/integration-tests/filepath_testutils.go [hunk 1]
```diff
@@ -1,32 +1,41 @@
...
+var testDataDir string
+var createTestDataDirOnce sync.Once
+
+func TestDataDir(t *testing.T) string {
+	createTestDataDirOnce.Do(func() {
+		var err error
+		_, filename, _, ok := runtime.Caller(0)
+		require.True(t, ok, "unable to get filename from runtime")
+		srcDir := filepath.Join(filepath.Dir(filename), "testdata")
+		testDataDir, err = ioutil.TempDir(baseTempDir, "testdata-")
```
[the remark this answers is not in the document]
Comment (adombeck) [reply pkg/runner/libfuzzer/integration-tests/filepath_testutils.go]: Done

**Comment by fmeum** [comment]
If possible, I would find the following ideal:
* Run the tests sequentially in CI, where the integration tests run longer anyway.
* Make it easy to run the tests in parallel locally, where they currently consume a good chunk of the time spent on running `go test ./...`.

**Comment by adombeck** [comment]
> If possible, I would find the following ideal:
> 
>     * Run the tests sequentially in CI, where the integration tests run longer anyway.
> 
>     * Make it easy to run the tests in parallel locally, where they currently consume a good chunk of the time spent on running `go test ./...`.

I think we would achieve that via
```
if os.Getenv("CI") != "" {
    t.Parallel()
}
```

Would you both be fine with that, @jochil @fmeum?

**Comment by fmeum** [comment]
Looks good to me!

**Comment by adombeck** [comment]
Note that the fix for a remaining issue increased the run time on system from ~5s to ~10s

**Comment by jochil** [comment]
I am not a fan of tests behaving differently depending on the environment and I think the "performance" gain is not worth it.
But if we really want to do it, I would prefer doing it in the Makefile by setting `go test -parallel=1`. Locally you just can run it with `go test ./...` or by adding a new make target `make test/parallel`

**Comment by fmeum** [comment]
> Note that the fix for a remaining issue increased the run time on my system from ~5s to ~10s

Can't we copy the build directory to prevent the locking?

**Comment by fmeum** [comment]
> I am not a fan of tests behaving differently depending on the environment and I think the "performance" gain is not worth it.
> But if we really want to do it, I would prefer doing it in the Makefile by setting `go test -parallel=1`. Locally you just can run it with `go test ./...` or by adding a new make target `make test/parallel`

That's also fine with me. The CI could explicitly run `-parallel=1` instead of relying on this variable.

I frequently run the tests locally while making edits and do think that the performance gain would be very helpful for my feedback loop.

**Comment by adombeck** [comment]
> Can't we copy the build directory to prevent the locking?

Done, run time is at ~6s now

**Comment by adombeck** [comment]
I added a new make target `test/parallel/sequential` which is now used in the CI. Everyone happy now, @jochil @fmeum? :)

**Comment by fmeum** [comment]
@adombeck Looks good!

**Review by jochil — APPROVED** [review]

**Pull request merged**